extensions/lang-js/src/main/java/io/crate/operation/language/PolyglotValues.java

```diff
@@ -41,26 +42,58 @@ class PolyglotValues {
 
     private static final TypeLiteral<Number> NUMBER_TYPE_LITERAL = new TypeLiteral<>() {
     };
-    private static final TypeLiteral<Map<?, ?>> MAP_TYPE_LITERAL = new TypeLiteral<>() {
+    private static final TypeLiteral<Object> OBJECT_LITERAL = new TypeLiteral<>() {
     };
 
     private PolyglotValues() {
     }
 
+    private static Object toCrateObject(Value value) {
+        if (value.isNull()) {
+            return null;
+        }
+        if (value.isHostObject()) {
+            return value.asHostObject();
+        }
+        if (value.hasIterator()) {
+            ArrayList<Object> items = new ArrayList<>();
+            Value iterator = value.getIterator();
+            while (iterator.hasIteratorNextElement()) {
+                items.add(toCrateObject(iterator.getIteratorNextElement()));
+            }
+            return items;
+        }
+        return value.as(OBJECT_LITERAL);
```

Comment:
At this point we know that `Value` is a non-host objects (because we didn't take a shortcut). 
Is it safe? 


On the other hand, if it's safe for non-host objects, using 
```
public static Object toCrateObject(Value value, DataType<?> type) {
    return type.implicitCast(value.as(OBJECT_LITERAL)); // takes care of NULL and internal Java type
}
```
should be safe for the same reason as it's safe here?

Reply:
> Is it safe?

It should be safe for most of the literals, except anything that is a collection.

I added a fixup that adds map handling here and more tests.